BZ1871254: Add section for RAID storage (#39515)
openshift/openshift-docs: modules/installation-special-config-mirrored-disk.adoc

This command produces a combined Ignition config in `$HOME/clusterconfig/worker-raid.ign`.
+
. Continue with the remainder of the {product-title} deployment, using the combined Ignition config to provision nodes.

== Configuring a RAID-enabled data volume

You can enable software RAID partitioning to provide an external data volume.

. Create a machine config that configures a data volume by using software RAID.

* To configure a data volume with RAID 1 on secondary disks, save the following machine config sample to a file in the `<installation_directory>/openshift` directory. For example, name the file `raidl-alt-storage.yaml`:
+
.RAID 1 on secondary disks
[source, yaml]
----
variant: openshift
version: 4.7.0
metadata:

>   name: raidl-alt-storage

Reviewer: ```suggestion
  name: raid1-alt-storage
```

  labels:
    machineconfiguration.openshift.io/role: worker
storage:
  disks:
    - device: /dev/sdc
      wipeTable: true
      partitions:
        - label: data-1
    - device: /dev/sdd
      wipeTable: true
      partitions:
        - label: data-2
  raid:
    - name: md-var-lib-containers
      level: raid1
      devices:
        - /dev/disk/by-partlabel/data-1
        - /dev/disk/by-partlabel/data-2
  filesystems:
    - device: /dev/md/md-var-lib-containers
      path: /var/lib/containers
      format: xfs
      wipeFilesystem: true
systemd:
  units:
    - contents: |
        [Unit]
        Before=local-fs.target
        Requires=systemd-fsck@dev-md-md\x2dvar\x2dlib\x2dcontainers.service
        After=systemd-fsck@dev-md-md\x2dvar\x2dlib\x2dcontainers.service

        [Mount]
        Where=/var/lib/containers
        What=/dev/md/md-var-lib-containers
        Type=xfs

        [Install]
        RequiredBy=local-fs.target
      enabled: true
      name: var-lib-containers.mount
----
. Apply the config to your cluster by running the following command:
+
[source, terminal]
----
$ oc create -f raidl-alt-storage.yaml
----

> . Save the machine config in case you need to update it in the future. 

Reviewer: This line is meant for Butane configs, which aren't stored in the cluster.  I think we can drop it when writing a MachineConfig directly.

* To configure a data volume with RAID 1 on secondary disks, save the following machine config sample to a file in the `<installation_directory>/openshift` directory. For example, name the file `raid1-alt-storage.yaml`:
+
.RAID 1 on secondary disks
[source, yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: raid1-alt-storage
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      disks:
      - device: /dev/sdc
        partitions:
        - label: data-1
        wipeTable: true
      - device: /dev/sdd
        partitions:
        - label: data-2
        wipeTable: true
      filesystems:
      - device: /dev/md/md-var-lib-containers
        format: xfs
        path: /var/lib/containers
        wipeFilesystem: true
      raid:
      - devices:
        - /dev/disk/by-partlabel/data-1
        - /dev/disk/by-partlabel/data-2
        level: raid1
        name: md-var-lib-containers
    systemd:
      units:
      - contents: |-
          [Unit]
          Before=local-fs.target
          Requires=systemd-fsck@dev-md-md\x2dvar\x2dlib\x2dcontainers.service

>           After=systemd-fsck@dev-md-md\x2dvar\x2dlib\x2dcontainers.service

Reviewer: If the user changes the name of the RAID device, these need to be changed also, and they're kind of a pain with the escaping.  We can address this by avoiding implying a mountpoint path in the device name, so that users aren't tempted to change it.  Let's rename the RAID from `md-var-lib-containers` to `md-data` and update these two lines, the `What` line, and the filesystem `device` field.

          [Mount]
          Where=/var/lib/containers
          What=/dev/md/md-var-lib-containers
          Type=xfs

          [Install]
          RequiredBy=local-fs.target
        enabled: true

>         name: var-lib-containers.mount

Reviewer: If users edit the sample to pick a different mountpoint, they also need to update this unit name or the unit won't activate.  We've had bug reports about this in the past.  We should probably add a comment saying that the unit name must correspond to the mountpoint, and can be generated with something like `echo $(systemd-escape -p $mountpoint).mount`.

This command produces a combined Ignition config in `$HOME/clusterconfig/worker-raid.ign`.
+
. Continue with the remainder of the {product-title} deployment, using the combined Ignition config to provision nodes.

> == Configuring a RAID-enabled data volume

Reviewer 2: This heading needs a unique ID.

You can enable software RAID partitioning to provide an external data volume.

> 

Reviewer 2: Needs a `.Procedure` heading

> . Create a machine config that configures a data volume by using software RAID.
> 
> * To configure a data volume with RAID 1 on secondary disks, save the following machine config sample to a file in the `<installation_directory>/openshift` directory. For example, name the file `raid1-alt-storage.yaml`:
> +
> .RAID 1 on secondary disks
> [source, yaml]
> ----

Reviewer 2: Rather than making this a bullet, I'd make it a continuation (on a new line) of the step it's under. A lot of times in our docs, we do something like this:
```suggestion
. Create a machine config that configures a data volume by using software RAID in the `<installation_directory>/openshift` directory.
+
.Sample RAID 1 on secondary disks configuration file named `raid1-alt-storage.yaml`
[source, yaml]
----
```
(My example file .header could probably stand some improvement, but just trying to illustrate :slightly_smiling_face: You can see lots of examples if you search `.Sample` in the repo.)

        partitions:
        - label: data-2
        wipeTable: true
      filesystems:
      - device: /dev/md/md-data
        format: xfs
        path: /var/lib/containers
        wipeFilesystem: true
      raid:
      - devices:
        - /dev/disk/by-partlabel/data-1
        - /dev/disk/by-partlabel/data-2
        level: raid1
        name: md-data
    systemd:
      units:
      - contents: |-
          [Unit]
          Before=local-fs.target
          Requires=systemd-fsck@dev-md-md\x2ddata.service
          After=systemd-fsck@dev-md-md\x2ddata.service

          [Mount]
          Where=/var/lib/containers
          What=/dev/md/md-data
          Type=xfs

          [Install]
          RequiredBy=local-fs.target
        enabled: true
        name: var-lib-containers.mount <1>
----
+
<1> If you choose a different mount point, you must update the unit name to correspond to your mount point. Otherwise the unit will not activate. You can generate a matching unit name with `echo $(systemd-escape -p $mountpoint).mount` where `$mountpoint` is your chosen mount point.
+
. Apply the config to your cluster by running the following command:
+
[source, terminal]
----

> $ oc create -f raid1-alt-storage.yaml

Reviewer 2: You might want to replace `raid1-alt-storage.yaml` with a user-replaceable variable such as `<raid_config_file>`. Sometimes users just copy and run code snippets without changing things, so it's best to make it super clear it's an example. If you want to use your example from above, you could put something like this under it:
> where `<raid_config_file>` is your RAID configuration file, for example `raid1-alt-storage.yaml`.

[source, yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: raid1-alt-storage
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      disks:
      - device: /dev/sdc
        partitions:
        - label: data-1
        wipeTable: true
      - device: /dev/sdd
        partitions:
        - label: data-2
        wipeTable: true
      filesystems:
      - device: /dev/md/md-data
        format: xfs
        path: /var/lib/containers
        wipeFilesystem: true
      raid:
      - devices:
        - /dev/disk/by-partlabel/data-1
        - /dev/disk/by-partlabel/data-2
        level: raid1
        name: md-data
    systemd:
      units:
      - contents: |-
          [Unit]
          Before=local-fs.target
          Requires=systemd-fsck@dev-md-md\x2ddata.service
          After=systemd-fsck@dev-md-md\x2ddata.service

          [Mount]
          Where=/var/lib/containers
          What=/dev/md/md-data
          Type=xfs

          [Install]
          RequiredBy=local-fs.target
        enabled: true
        name: var-lib-containers.mount
----

[IMPORTANT]
====

> If you choose a different mount point, you must update your `raid1-alt-storage.yaml` file with a unit name that corresponds to this mount point. Otherwise the unit will not activate. You can generate a matching unit name with `echo $(systemd-escape -p $mountpoint).mount` where `$mountpoint` is your chosen mount point.

Reviewer: It'd be good to explicitly say here that the unit named `var-lib-containers.mount` needs to be changed.  In similar situations, we've attached a footnote to the specific config line that needs extra care.

Author: Gotcha, I'll update this to a specific line callout
